### ISSUE TITLE
running the tests deletes the default user

### ISSUE THREAD

**Issue by ltalirz** [aiidateam/aiida_core#2407]
I encountered the following error while trying to run my first WorkChain on the new `provenance_redesign` branch, python 2.7.15:

```
  File "/Users/leopold/Personal/Postdoc-MARVEL/repos/aiida/aiida_rmq/aiida/orm/implementation/django/node.py", line 127, in __init__
    user = orm.User.objects.get_default().backend_entity
AttributeError: 'NoneType' object has no attribute 'backend_entity'
```
This is in a profile that was migrated to the current `provenance_redesign` branch (not created from scratch).

To me, this raises a few questions:

 * how is it possible that my profile does not have a default user?
 * should a profile with no default user be allowed? I.e. should `orm.User.objects.get_default()` be allowed to return `None` or should it raise an exception?
 * if you don't have a default user, how do you add one? I tried adding a new user via `verdi user` (which worked) but, apparently, this is not enough. How do I make it the default one?

Perhaps @giovannipizzi could comment?

**Comment by giovannipizzi** [comment]
I think you copied only part of the error.
Could you tell us if there is any mention of the default user in the config file (.aiida/config.json)? (`default_user_email` inside a profile section)

What does `verdi profile list` say about the default user?

Probably it's possible that there is no way of setting a default user and a new command should be added to `verdi user` for simplicity (even if I think that probably one could achieve what you want to do with `verdi setup` but I didn't try).

**Comment by ltalirz** [comment]
> I think you copied only part of the error.

thanks for pointing this out - fixed!

> Could you tell us if there is any mention of the default user in the config file (.aiida/config.json)? 

Indeed, there is a default user specified but it doesn't show up in `verdi user list`

> What does verdi profile list say about the default user?

I guess you mean `verdi profile show`. It correctly shows the information stored in the config file (but the user doesn't seem to exist in the DB).

> Probably it's possible that there is no way of setting a default user and a new command should be added to verdi user for simplicity 

Either to verdi user or, perhaps, rather to `verdi profile`, since one is actually setting the default user for a profile.

**Comment by ltalirz** [comment]
> so the problem is that the user is not in the DB, right? Do you know how this happened?

I don't really know - but it might be my fault in some way. Let's for the moment assume it is my fault. 

I would still say that we need to decide whether having no default user is allowed or not. Currently, part of the codebase seems to assume that it cannot happen, which would mean we should raise an exception, ideally with instructions on how to make a user the default one.

**Comment by ltalirz** [comment]
Just mentioning that @waychal had exactly the same problem (no default user, no users showing up in `verdi user list`).

Furthermore, I had the user that I previously configured via `verdi user configure` disappear from `verdi user list` after rebooting my system.

**Comment by ltalirz** [comment]
Just happened again to me as well: the users seem to be randomly deleted - i.e. I'm still on the same profile but now `verdi user list` is empty.

**Comment by giovannipizzi** [comment]
That's really weird...
could you check if there are places in the code where a delete call on the user is called, to see if it's inadvertently triggered as a side-effect of some other call?

**Comment by ltalirz** [comment]
Not sure whether this is related but on PR  #2421 I'm getting an error (on python2.7 only)
`DETAIL:  Key (user_id)=(21) is not present in table "db_dbuser".`:
https://travis-ci.org/aiidateam/aiida_core/jobs/486559586#L3074

**Comment by ltalirz** [comment]
One of the problems might have been that I was running on a test database (and might have been running tests).
Still, I have the feeling this did not happen in previous AiiDA versions (and it is annoying, since it prevents you from keeping to do tests on the test database)

**Comment by giovannipizzi** [comment]
I have the feeling this happened in your PR #2214 (but with the commit by @muhrin), where before there was something like
```python
         # We create the user only once:		
         # Otherwise, get_automatic_user() will fail when the		
         # user is recreated because it caches the user!		
         # In any case, store it in self.user though		
         try:		
             self.user = DbUser.objects.get(email=get_configured_user_email())		
         except ObjectDoesNotExist:		
             self.user = DbUser.objects.create_user(get_configured_user_email(), 'fakepwd')
```
and now
```
DbUser.objects.all().delete()  # pylint: disable=no-member
```

I think also the deleted comment is quite explicative. Is it possible? I don't know though if @muhrin moved the logic somewhere else

**Comment by ltalirz** [comment]
I think the first code you pasted is not related but the second one in `clean_db` is:
https://github.com/aiidateam/aiida_core/pull/2214/files#diff-d85c6625843bfeca5bb54998de77fe55L102

Before:
```python
DbUser.objects.exclude(email=get_configured_user_email()).delete()
```
After:
```python
DbUser.objects.all().delete() 
```
Indeed this would mean that the tests could still pass (since a default user will be created on demand), but running the tests will always result in deletion of the default user.

**Comment by ltalirz** [comment]
To summarize: 
 * yes, the current version of `provenance_redesign` results in all ORM objects being deleted after the tests are run
 * the issue observed in #2400 was different (see #2515)
 * as pointed out by @szoupanos , tests should not rely on anything being present in the database, not even a default user

I personally still find it annoying not to have a default user in the database because it means you can't do anything in the shell without setting up the user first.
Furthermore, I find it inconsistent that after creating a test database via `verdi quicksetup` there *will* be a default user that will be deleted the first time you run a test.

Anyhow, I can also see the argument by spyros. Feel free to close

**Comment by ltalirz** [comment]
To me, the larger question here is what a user is supposed to be able to do with a test profile.

If the policy is that you cannot do anything with a test profile other than running the tests, then I think there is actually no need to set it up as a regular profile using `verdi quicksetup`.
Instead we could use something like this
https://github.com/aiidateam/aiida_core/blob/da6bb4d1657cdf742adb4ef9fb1739e08404320d/aiida/backends/tests/manage/external/test_postgres.py#L33-L36
and create the temporary profile on the fly.

**Comment by ltalirz** [comment]
this last comment will be addressed in #2771, so this can be closed
